### DISCUSSION TITLE
How to update position of popover dialog

### DISCUSSION THREAD

**Discussion by martpet** [adobe/react-spectrum#3676] [Q&A] [answered]
I'm trying to use `useOverlayPosition` to update the position of the popover, but no success. I assume `useOverlayPosition` is meant  to be used only with custom overlays. 

Is there any way to update the position, besides closing and reopening the dialog?

    <DialogTrigger type="popover">
      <ActionButton />
      <Dialog />
    </DialogTrigger>

**Comment by snowystinger** [comment, marked as answer]
You're using the method returned by the hook, updatePosition? Can you show an example of what you're trying to do?

**Reply by martpet** [reply]
I'm naively trying this:

```
const overlayRef = useRef(null);
const targetRef = useRef(null);
const { updatePosition } = useOverlayPosition({ overlayRef, targetRef });
const { toolbarPosition } = useToolbarPosition();
const dispatch = useDispatch();

useEffect(() => {
  updatePosition();
}, [toolbarPosition]);

return (
  <DialogTrigger type="popover">
    <ActionButton ref={targetRef}>Foo</ActionButton>
    <Dialog ref={overlayRef}>
      <>
        <Button
          variant="secondary"
          onPress={() => dispatch(toolbarPositionChanged('left'))}
        >
          left
        </Button>
        <Button
          variant="secondary"
          onPress={() => dispatch(toolbarPositionChanged('top'))}
        >
          top
        </Button>
      </>
    </Dialog>
  </DialogTrigger>
);
```

**Reply by snowystinger** [reply]
I think you meant to pass the position into the useOverlayPosition hook?
```
let [position, setPosition] = useState('left')
const { updatePosition } = useOverlayPosition({ overlayRef, targetRef, position });
```

**Reply by martpet** [reply]
Actually "toolbarPosition" is for a container (a toolbar) that wraps the ActionButton - the trigger of the popover.

Here is an example: https://test.trip.pictures
Click the Settings icon and then click "Layout"  and change the position of the toolbar.

I want to move the popover next to its trigger when the toolbar position changes

**Reply by snowystinger** [reply]
I think you're probably pretty close, but here is how you can update in response to a layout change https://codesandbox.io/s/stoic-northcutt-lgvtbi?file=/src/App.js

**Reply by martpet** [reply]
Thanks for your example. I might use it in the future, but for now I prefer not to create a custom dialog component. I will stick with the built-in features of `<DialogTrigger>`. Hopefully, one day, there will be an option to reposition a `<Dialog>` inside a `<DialogTrigger type="popover">`.

It's a great library!

**Reply by snowystinger** [reply]
I see, I thought you were using our hooks.
I think for our components, we wouldn't encourage a popover moving around the screen in response to a layout shift. That's really hard for a person with Glaucoma who may have a blind spot where the popover moves to, and they'd just be left staring at a blank spot in the screen where they were working on something a moment ago. This might be ok for some users if the action they took was to redo the layout. However, if they were using a screen reader for assistance, they may have a hard time realizing they are still inside the popover and need to close it out.

That's not to say never, but we'd have to consider it pretty carefully
